File: regression/rt_mhd_brio_wu.c
#include <math.h>
#include <stdio.h>

#include <gkyl_alloc.h>
#include <gkyl_moment.h>
#include <gkyl_util.h>
#include <gkyl_wv_mhd.h>
#include <rt_arg_parse.h>

struct mhd_ctx {
  double gas_gamma; // gas constant
};

void
evalMhdInit(double t, const double* GKYL_RESTRICT xn, double* GKYL_RESTRICT fout, void *ctx)
{
  struct mhd_ctx *app = ctx;
  double x = xn[0];
  double gas_gamma = app->gas_gamma;

  double bx = 0.75;
  double rhol = 1.0, rhor = 0.125;
  double byl = 1.0, byr = -1.0;
  double pl = 1.0, pr = 0.1;

  double rho = rhor, by = byr, p = pr;
  if (x<0.5) {
    rho = rhol;
    by = byl;
    p = pl;
  }

  fout[0] = rho;
  fout[1] = 0.0; fout[2] = 0.0; fout[3] = 0.0;
  fout[4] = p/(gas_gamma-1) + 0.5*(bx*bx + by*by);
  fout[5] = bx; fout[6] = by; fout[7] = 0.0;
}

struct mhd_ctx
mhd_ctx(void)
{
  return (struct mhd_ctx) { .gas_gamma = 2.0 };
}

int
main(int argc, char **argv)
{
  struct gkyl_app_args app_args = parse_app_args(argc, argv);

  int NX = APP_ARGS_CHOOSE(app_args.xcells[0], 400);

  if (app_args.trace_mem) {
    gkyl_cu_dev_mem_debug_set(true);
    gkyl_mem_debug_set(true);
  }
  struct mhd_ctx ctx = mhd_ctx(); // context for init functions

  // equation object
  const struct gkyl_wv_mhd_inp inp = {
    .gas_gamma = ctx.gas_gamma,
    .divergence_constraint = GKYL_MHD_DIVB_NONE,
  };
  struct gkyl_wv_eqn *mhd = gkyl_wv_mhd_new(&inp);

  struct gkyl_moment_species fluid = {
    .name = "mhd",

    .equation = mhd,
    .evolve = 1,
    .ctx = &ctx,
    .init = evalMhdInit,

    .bcx = { GKYL_SPECIES_COPY, GKYL_SPECIES_COPY },
  };
  
  // VM app
  struct gkyl_moment app_inp = {
    .name = "mhd_brio_wu",

    .ndim = 1,
    .lower = { 0.0 },
    .upper = { 1.0 }, 
    .cells = { NX },

    .num_species = 1,
    .species = { fluid },
  };

  // create app object
  gkyl_moment_app *app = gkyl_moment_app_new(&app_inp);

  // start, end and initial time-step
  double tcurr = 0.0, tend = 0.1;

  // initialize simulation
  gkyl_moment_app_apply_ic(app, tcurr);
  gkyl_moment_app_write(app, tcurr, 0);

  // compute estimate of maximum stable time-step
  double dt = gkyl_moment_app_max_dt(app);

  long step = 1, num_steps = app_args.num_steps;
  while ((tcurr < tend) && (step <= num_steps)) {
    printf("Taking time-step %ld at t = %g ...", step, tcurr);
    struct gkyl_update_status status = gkyl_moment_update(app, dt);
    printf(" dt = %g\n", status.dt_actual);
    
    if (!status.success) {
      printf("** Update method failed! Aborting simulation ....\n");
      break;
    }
    tcurr += status.dt_actual;
    dt = status.dt_suggested;

    step += 1;
  }

  gkyl_moment_app_write(app, tcurr, 1);
  gkyl_moment_app_stat_write(app);

  struct gkyl_moment_stat stat = gkyl_moment_app_stat(app);

  // simulation complete, free resources
  gkyl_wv_eqn_release(mhd);
  gkyl_moment_app_release(app);

  printf("\n");
  printf("Number of update calls %ld\n", stat.nup);
  printf("Number of failed time-steps %ld\n", stat.nfail);
  printf("Species updates took %g secs\n", stat.species_tm);
  printf("Field updates took %g secs\n", stat.field_tm);
  printf("Total updates took %g secs\n", stat.total_tm);
  
  return 0;
}

File: regression/lua/rt_mhd_brio_wu.lua
local Moments = G0.Moments
local MHD = G0.Moments.Eq.MHD

-- Physical constants (using normalized code units).
gas_gamma = 2.0 -- Adiabatic index.

rhol = 1.0 -- Left fluid mass density.
ul = 0.0 -- Left fluid velocity.
pl = 1.0 -- Left fluid pressure.
Bx_l = 0.75 -- Left magnetic field (x-direction).
By_l = 1.0 -- Left magnetic field (y-direction).

rhor = 0.125 -- Right fluid mass density.
ur = 0.0 -- Right fluid velocity.
pr = 0.1 -- Right fluid pressure.
Bx_r = 0.75 -- Right magnetic field (x-direction).
By_r = -1.0 -- Right magnetic field (y-direction).

-- Simulation parameters.
Nx = 400 -- Cell count (x-direction).
Lx = 1.0 -- Domain size (x-direction).
cfl_frac = 0.95 -- CFL coefficient.

t_end = 0.1 -- Final simulation time.
num_frames = 1 -- Number of output frames.
field_energy_calcs = GKYL_MAX_INT -- Number of times to calculate field energy.
integrated_mom_calcs = GKYL_MAX_INT -- Number of times to calculate integrated moments.
dt_failure_tol = 1.0e-4 -- Minimum allowable fraction of initial time-step.
num_failures_max = 20 -- Maximum allowable number of consecutive small time-steps.

momentApp = Moments.App.new {
  
  tEnd = t_end,
  nFrame = num_frames,
  fieldEnergyCalcs = field_energy_calcs,
  integratedMomentCalcs = integrated_mom_calcs,
  dtFailureTol = dt_failure_tol,
  numFailuresMax = num_failures_max,
  lower = { 0.0 },
  upper = { Lx },
  cells = { Nx },
  cflFrac = cfl_frac,
    
  -- Boundary conditions for configuration space.
  periodicDirs = { }, -- Periodic directions (none).
  
  -- Fluid.
  fluid = Moments.Species.new {
    equation = MHD.new {
      gasGamma = gas_gamma,
      divergenceConstraint = G0.DivB.None
    },
    
    -- Initial conditions function.
    init = function (t, xn)
      local x = xn[1]

      local rho = 0.0
      local u = 0.0
      local p = 0.0

      local Bx = 0.0
      local By = 0.0
      local Bz = 0.0 -- Magnetic field (x-direction).

      if x < 0.5 then
        rho = rhol -- Fluid mass density (left).
        u = ul -- Fluid velocity (left).
        p = pl -- Fluid pressure (left).

        Bx = Bx_l -- Magnetic field (x-direction, left).
        By = By_l -- Magnetic field (y-direction, left).
      else
        rho = rhor -- Fluid mass density (right).
        u = ur -- Fluid velocity (right).
        p = pr -- Fluid pressure (right).

        Bx = Bx_r -- Magnetic field (x-direction, right).
        By = By_r -- Magnetic field (y-direction, right).
      end
  
      local mom_x = rho * u -- Fluid momentum density (x-direction).
      local mom_y = 0.0 -- Fluid momentum density (y-direction).
      local mom_z = 0.0 -- Fluid momentum density (z-direction).
      local Etot = (p / (gas_gamma - 1.0)) + (0.5 * rho * u * u) + (0.5 * ((Bx * Bx) + (By * By) + (Bz * Bz))) -- Fluid total energy density.
      
      return rho, mom_x, mom_y, mom_z, Etot, Bx, By, Bz, 0.0
    end,
  
    evolve = true, -- Evolve species?
    bcx = { G0.SpeciesBc.bcCopy, G0.SpeciesBc.bcCopy } -- Copy boundary conditions (x-direction).
  }
}

-- Run application.
momentApp:run()